pkg/rbac/cacheValidation.go

```diff
@@ -0,0 +1,213 @@
+
+	// Watch CLUSTERROLES
+
+	// Watch ROLEBINDINGS
+
+	// Watch CLUSTERRROLEBINDINGS
+
+	// Watch GROUPS
+
+	// Watch CRDS
+}
+
+// Start watching for changes to a resource and trigger the action to update the cache.
+func (w watchResource) start(ctx context.Context) {
+	for {
+		watch, watchError := w.dynamicClient.Resource(w.gvr).Watch(ctx, metav1.ListOptions{})
+		if watchError != nil {
+			klog.Warningf("Error watching %s, waiting 5 seconds before retry. Error: %s", w.gvr.String(), watchError)
+			time.Sleep(5 * time.Second) // Wait before retrying.
+			continue
+		}
+
+		defer watch.Stop()
+		klog.V(2).Infof("Watching resource: %s", w.gvr.String())
+
+		for {
+			select {
+			case <-ctx.Done():
+				klog.V(2).Info("Stopped watching resource. ", w.gvr.String())
+				watch.Stop()
+				return
+
+			case event := <-watch.ResultChan(): // Read events from the watch channel.
+				klog.V(6).Infof("Event: %s \tResource: %s  ", event.Type, w.gvr.String())
+				o, err := runtime.UnstructuredConverter.ToUnstructured(runtime.DefaultUnstructuredConverter, &event.Object)
+				if err != nil {
+					klog.Warningf("Error converting %s event.Object to unstructured.Unstructured. Error: %s",
+						w.gvr.Resource, err)
+				}
+				obj := &unstructured.Unstructured{Object: o}
+
+				switch event.Type {
+				case "ADDED":
+					if w.onAdd != nil {
+						w.onAdd(obj)
+					}
+				case "MODIFIED":
+					if w.onModify != nil {
+						w.onModify(obj)
+					}
+				case "DELETED":
+					if w.onDelete != nil {
+						w.onDelete(obj)
+					}
+
+				default:
+					klog.V(2).Infof("Unexpected event, waiting 5 seconds and restarting watch for %s", w.gvr.String())
+					watch.Stop()
+					time.Sleep(5 * time.Second)
+				}
+			}
+		}
+	}
+}
+
+// Update the cache when a namespace is ADDED.
+func (c *Cache) namespaceAdded(obj *unstructured.Unstructured) {
+	// Addd namespace to shared cache.
+	c.shared.nsCache.lock.Lock()
+	c.shared.namespaces = append(c.shared.namespaces, obj.GetName())
+	c.shared.nsCache.updatedAt = time.Now()
+	c.shared.nsCache.lock.Unlock()
+
+	// Add the namespace to each user in UserData cache.
+	c.usersLock.Lock()
+	defer c.usersLock.Unlock()
+	wg := sync.WaitGroup{}
+	lock := sync.Mutex{}
+	for _, userCache := range c.users {
+		wg.Add(1)
+		go func(userCache *UserDataCache) { // All users updated asynchhronously
+			defer wg.Done()
+			userCache.getSSRRforNamespace(context.TODO(), c, obj.GetName(), &lock)
+		}(userCache)
+	}
+	wg.Wait() // Wait until all users have been updated.
+
+	// Note: The ManagedCluster and DisabledClusters cache will get updated
+	// when we receive the ManagedCluster ADD event.
+}
+
+// Update the cache when a namespace is DELETED.
+func (c *Cache) namespaceDeleted(obj *unstructured.Unstructured) {
+	// Delete from Namespaces shared cache.
+	c.shared.nsCache.lock.Lock()
+	defer c.shared.nsCache.lock.Unlock()
+	ns := obj.GetName()
+	newNamespaces := make([]string, 0)
+	for _, n := range c.shared.namespaces {
```

Comment:
Is it possible to convert namespaces also to a map structure so that we don't have loop through it to delete?

Reply:
Yes! I want to change it also for consistency with the other data, however that's a big change, so I'll keep it separate from this PR.